msgraph/directory_objects_test.go

```diff
@@ -0,0 +1,190 @@
+	"github.com/manicminer/hamilton/internal/utils"
+	"github.com/manicminer/hamilton/msgraph"
+	"github.com/manicminer/hamilton/odata"
+)
+
+type DirectoryObjectsClientTest struct {
+	connection   *test.Connection
+	client       *msgraph.DirectoryObjectsClient
+	randomString string
+}
+
+func TestDirectoryObjectsClient(t *testing.T) {
+	rs := test.RandomString()
+	c := DirectoryObjectsClientTest{
+		connection:   test.NewConnection(auth.MsGraph, auth.TokenVersion2),
+		randomString: rs,
+	}
+	c.client = msgraph.NewDirectoryObjectsClient(c.connection.AuthConfig.TenantID)
+	c.client.BaseClient.Authorizer = c.connection.Authorizer
+
+	g := GroupsClientTest{
+		connection:   test.NewConnection(auth.MsGraph, auth.TokenVersion2),
+		randomString: rs,
+	}
+	g.client = msgraph.NewGroupsClient(c.connection.AuthConfig.TenantID)
+	g.client.BaseClient.Authorizer = c.connection.Authorizer
+
+	u := UsersClientTest{
+		connection:   test.NewConnection(auth.MsGraph, auth.TokenVersion2),
+		randomString: rs,
+	}
+	u.client = msgraph.NewUsersClient(c.connection.AuthConfig.TenantID)
+	u.client.BaseClient.Authorizer = c.connection.Authorizer
+
+	user := testUsersClient_Create(t, u, msgraph.User{
+		AccountEnabled:    utils.BoolPtr(true),
+		DisplayName:       utils.StringPtr("test-user"),
+		MailNickname:      utils.StringPtr(fmt.Sprintf("test-user-directoryobject-%s", c.randomString)),
+		UserPrincipalName: utils.StringPtr(fmt.Sprintf("test-user-directoryobject-%s@%s", c.randomString, c.connection.DomainName)),
+		PasswordProfile: &msgraph.UserPasswordProfile{
+			Password: utils.StringPtr(fmt.Sprintf("IrPa55w0rd%s", c.randomString)),
+		},
+	})
+
+	newGroup1 := msgraph.Group{
+		DisplayName:     utils.StringPtr("test-group-directoryobject-member"),
+		MailEnabled:     utils.BoolPtr(false),
+		MailNickname:    utils.StringPtr(fmt.Sprintf("test-group-directoryobject-member-%s", c.randomString)),
+		SecurityEnabled: utils.BoolPtr(true),
+	}
+	newGroup1.Members = &msgraph.Members{*user.DirectoryObject}
+	group1 := testGroupsClient_Create(t, g, newGroup1)
+
+	newGroup2 := msgraph.Group{
+		DisplayName:     utils.StringPtr("test-group-directoryobject"),
+		MailEnabled:     utils.BoolPtr(false),
+		MailNickname:    utils.StringPtr(fmt.Sprintf("test-group-directoryobject-%s", c.randomString)),
+		SecurityEnabled: utils.BoolPtr(true),
+		Members: &msgraph.Members{
+			*group1.DirectoryObject,
+			*user.DirectoryObject,
+		},
+	}
+	group2 := testGroupsClient_Create(t, g, newGroup2)
+
+	testDirectoryObjectsClient_Get(t, c, *user.ID)
+	testDirectoryObjectsClient_Get(t, c, *group1.ID)
+	testDirectoryObjectsClient_GetMemberGroups(t, c, *user.ID, true, []string{*group1.ID, *group2.ID})
+	testDirectoryObjectsClient_GetMemberObjects(t, c, *group1.ID, true, []string{*group2.ID})
+	testDirectoryObjectsClient_GetByIds(t, c, []string{*group1.ID, *group2.ID, *user.ID}, []string{odata.ShortTypeGroup})
+	testDirectoryObjectsClient_Delete(t, c, *group1.ID)
```

Comment:
using something like [Testify's `suite` package](https://pkg.go.dev/github.com/stretchr/testify/suite?utm_source=godoc) would make this easier fwiw (since these'd be methods hanging off of that struct with a Setup, as your doing here?)

Reply:
cool, will take a look at that and improving the test setup sprawl 👍